README.md
# Semantic Image Segmentation
**Members** : <a href="https://github.com/vegovs">Vegard Bergsvik Øvstegård</a>

**Supervisors** : <a href="https://www.mn.uio.no/ifi/personer/vit/jimtoer/">Jim Tørresen</a>

## Status
![Python package](https://github.com/gil-uav/semantic-image-segmentation/workflows/Python%20package/badge.svg?branch=master) ![Python Package using Conda](https://github.com/gil-uav/semantic-image-segmentation/workflows/Python%20Package%20using%20Conda/badge.svg) [![Codacy Badge](https://app.codacy.com/project/badge/Grade/41c8bd8d0049413a9432bdd78e4e3869)](https://www.codacy.com/gh/gil-uav/semantic-image-segmentation/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=gil-uav/semantic-image-segmentation&amp;utm_campaign=Badge_Grade) [![DeepSource](https://deepsource.io/gh/gil-uav/semantic-image-segmentation.svg/?label=active+issues&show_trend=true)](https://deepsource.io/gh/gil-uav/semantic-image-segmentation/?ref=repository-badge) [![DeepSource](https://deepsource.io/gh/gil-uav/semantic-image-segmentation.svg/?label=resolved+issues&show_trend=true)](https://deepsource.io/gh/gil-uav/semantic-image-segmentation/?ref=repository-badge)

## Description

This repository aims to implement and produce trained networks in semantic image segmentation for
[orthopohots](https://www.sciencedirect.com/topics/earth-and-planetary-sciences/orthophoto).
Current network structure is [U-net](https://lmb.informatik.uni-freiburg.de/people/ronneber/u-net/).

## Dependencies
* [Python](https://www.python.org/) (version 3.6, 3.7 or 3.8)
* [Pip](https://virtualenv.pypa.io/en/latest/)
* [virtualenv](https://virtualenv.pypa.io/en/latest/) or:
* [conda](https://docs.conda.io/en/latest/)
* [Cuda](https://developer.nvidia.com/cuda-10.2-download-archive) version 10.2

## Installation

```console
git clone https://github.com/gil-uav/semantic-image-segmentation.git
```

#### virtualenv

```console
python3 -m venv env
source env/bin/activate
pip install -r requirements.txt
```

#### Conda
```console
conda env create --file environment.yml
conda activate seg
```

Uninstall Pillow and install Pillow-SIMD:
```console
pip uninstall pillow
pip install pillow-simd
```
If you have a AVX2-enabled CPU, check with `grep avx2 /proc/cpuinfo`, you can install Pillow-SIMD with:
```console
pip uninstall pillow
CC="cc -mavx2" pip install -U --force-reinstall pillow-simd
```
This should have slightly better performance than the SSE4(default) version and much better than the standard Pillow
 package.

NB! Remember to uninstall and reinstall Pillow-SIMD. In some cases, python might not find the PIL
package, however a reinstall fixes this 99% of the time.

## Usage

### Training
The application fetches some configurations and parameters from a .env file if it exists.
Run `python train.py --help` to see all other arguments. The package is using [pytorch-lighting](https://github.com/PyTorchLightning/pytorch-lightning) and inherits all its arguments.

The data is expected to be structured like this:
```
data/
    images/
    masks/
```
The path do data us set using --dp argument.

#### Console example
This example stores the checkpoints and logs under the default_root_dir, uses all available GPUs and
fetches training data from --dp.

```console
python train.py --default_root_dir=/shared/use/this/ --gpus=-1 --dp=/data/is/here/
```

#### .env example
Only these arguments are fetched from .env, the rest must be passed through the CLI.
```
# Model config
N_CHANNELS=3
N_CLASSES=1
BILINEAR=True

# Hyperparameters
EPOCHS=300 # Epochs
BATCH_SIZE=4 # Batch size
LRN_RATE=0.001 # Learning rate
VAL_PERC=15 # Validation percent
TEST_PERC=15 # Testing percent
IMG_SIZE=512  # Image size
VAL_INT=1 # Validation interval
ACC_GRAD=4 # Accumulated gradients, number = K.
GRAD_CLIP=1.0 # Clip gradients with norm above give value
EARLY_STOP=10 # Early stopping patience(Epochs)

# Other
PROD=False # Turn on or off debugging APIs
DIR_DATA="data/" # Where dataset is stored
DIR_ROOT_DIR="/shared/use/this/" # Where logs and checkpoint will be stored
WORKERS=4 # Number of workers for data- and validation loading
DISCORD_WH=httpsomethingwebhoowawnserisalways42
```

### Performance tips:
* Try with different number of workers, but more than 0. A good starting point
is `workers = cores * (threads per core)`.
* Install Pillow-SIMD as described in [Installation](#installation).

## Features
### ML:
* Distributed data parallel training
* Early stopping
* ADAM optimizer
* Gradient clipping
* ReduceLROnPlateau learning rate scheduler
* Logging to Tensorboard
* Metrics:
    * Loss
    * F1
    * Precision
    * Recall
    * Visualise images
* Gradient Accumulation(NB! Might conflict with batch-norm!)

### Ease of use:
* Add hyper-parameters and arguments from console
* Load hyper-parameters from .env
* Training finished notification to Discord

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
[MIT Licence](https://github.com/gil-uav/semantic-image-segmentation/blob/master/LICENSE)


CHANGELOG.md
## 3.1.0 (2020-09-29)

### Feat

- **train**: discord notification on train finished

### Fix

- added test loop

## 3.0.0 (2020-09-28)

### Feat

- **unet_model**: added ReduceLROnPlateau
- **unet_model**: added lr-scheduler
- metric update and bug fixes
- **train.py**: added accumulative grad and grad clipping
- **train.py**: re-wrote entire training loop
- **pre_processing**: added rescaling switch
- **dataset**: added data augmentation switch

### Fix

- fixed env variable for directories

## 2.0.0 (2020-09-25)

### Refactor

- removed .env file
- **unet**: moved arguments
- **lightning**: refractored pytorch code to pytorch-lightning
- black refractor
- black refractor
- **pre_processing**: duplication fix and type checking

### BREAKING CHANGE

- Complete refractor.
- Complete re-write of main training-loop.
- Random- Brightness, Saturation and Contrast replaced by RandomColorJitter.

### Perf

- **train**: non blocking on images

## 1.0.0 (2020-09-23)

### Perf

- **train.py**: enabled cudnn algorithm optimization

### Feat

- **tensorboard**: added validation loss
- **train.py**: added graph logging to tensorboard
- initial commit


eval.py
import torch
from torch.cuda.amp import autocast
from tqdm import tqdm
from pytorch_lightning.metrics.functional import (
    accuracy,
    precision,
    iou,
    recall,
    f1_score,
)


def eval_net(net, loader, device, criterion):
    """
    Evaluates network with accuracy, F1, precision, recall and IoU.

    Parameters
    ----------
    net : nn.Module
        Netword to evaluate.
    loader : nn.Dataloader
        Data to evaluate network on.
    device : torch.Device
        CPU or GPU
    criterion : nn._Loss
        Loss function

    Returns
    -------
    {"Accuracy": _accuracy / val_amount, "F1": _f1 / val_amount, "Precision": _precision / val_amount,
    "Recall": _recall / val_amount, "IOU": _iou / val_amount}
    """
    net.eval()  # Turn off training layers
    mask_type = torch.float32 if net.module.n_classes == 1 else torch.long
    val_amount = len(loader)  # the number of batch
    _accuracy = _f1 = _precision = _recall = _iou = _loss = 0

    with tqdm(
        total=val_amount, desc="Validating", unit="batch", position=0, leave=True
    ) as p_bar:
        for batch in loader:
            images, masks = batch["image"], batch["mask"]
            images = images.to(device=device, dtype=torch.float32)
            masks = masks.to(device=device, dtype=mask_type)

            # Automatic mixed precision(Hopefully to float16 to utilize tensor cores)
            with autocast(enabled=True):
                with torch.no_grad():  # Disable gradient calculation
                    masks_pred = net(images)
                    _loss += criterion(masks_pred, masks).item()

            pred = torch.sigmoid(masks_pred)
            pred = (pred > 0.5).float()

            # Calculate metrics
            num_classes = net.module.n_classes + 1
            _accuracy += accuracy(pred, masks, num_classes=num_classes).item()
            _f1 += f1_score(pred, masks, num_classes=num_classes).item()
            _precision += precision(pred, masks, num_classes=num_classes).item()
            _recall += recall(pred, masks, num_classes=num_classes).item()
            _iou += iou(pred, masks, num_classes=num_classes).item()

            p_bar.set_postfix(
                **{
                    "L": _loss / val_amount,
                    "A": _accuracy / val_amount,
                    "F1": _f1 / val_amount,
                    "P": _precision / val_amount,
                    "R": _recall / val_amount,
                    "IOU": _iou / val_amount,
                }
            )
            p_bar.update()

    net.train()  # Turn on training layers
    return {
        "Avg Loss": _loss / val_amount,
        "Accuracy": _accuracy / val_amount,
        "F1": _f1 / val_amount,
        "Precision": _precision / val_amount,
        "Recall": _recall / val_amount,
        "IOU": _iou / val_amount,
    }


train.py
import logging
import os
import sys
from argparse import ArgumentParser
from knockknock import discord_sender

import torch
from dotenv import load_dotenv
from pytorch_lightning import Trainer
from pytorch_lightning.callbacks import EarlyStopping, LearningRateLogger
from torch.backends import cudnn

from unet.unet_model import UNet

load_dotenv(verbose=True)


@discord_sender(webhook_url=os.getenv("DISCORD_WH"))
def main():
    parser = ArgumentParser()

    parser = UNet.add_model_specific_args(parser)
    parser = Trainer.add_argparse_args(parser)

    args = parser.parse_args()

    prod = bool(os.getenv("PROD"))
    if prod:
        logging.info("Training i production mode, disabling all debugging APIs")
        torch.autograd.set_detect_anomaly(False)
        torch.autograd.profiler.profile(enabled=False)
        torch.autograd.profiler.emit_nvtx(enabled=False)
    else:
        logging.info("Training i development mode, debugging APIs active.")
        torch.autograd.set_detect_anomaly(True)
        torch.autograd.profiler.profile(
            enabled=True, use_cuda=True, record_shapes=True, profile_memory=True
        )
        torch.autograd.profiler.emit_nvtx(enabled=True, record_shapes=True)

    model = UNet(**vars(args))

    logging.info(
        f"Network:\n"
        f"\t{model.hparams.n_channels} input channels\n"
        f"\t{model.hparams.n_classes} output channels (classes)\n"
        f'\t{"Bilinear" if model.hparams.bilinear else "Transposed conv"} upscaling'
    )

    cudnn.benchmark = True  # cudnn Autotuner
    cudnn.enabled = True  # look for optimal algorithms

    lr_monitor = LearningRateLogger(logging_interval="step")
    early_stopping = EarlyStopping(
        monitor="val_loss",
        patience=5 if not os.getenv("EARLY_STOP") else int(os.getenv("EARLY_STOP")),
        verbose=True,
    )

    try:
        trainer = Trainer.from_argparse_args(
            args,
            gpus=-1,
            precision=16,
            distributed_backend="ddp",
            callbacks=[lr_monitor],
            early_stop_callback=early_stopping,
            accumulate_grad_batches=1
            if not os.getenv("ACC_GRAD")
            else int(os.getenv("ACC_GRAD")),
            gradient_clip_val=0.0
            if not os.getenv("GRAD_CLIP")
            else float(os.getenv("GRAD_CLIP")),
            max_epochs=1000 if not os.getenv("EPOCHS") else int(os.getenv("EPOCHS")),
            default_root_dir=os.getcwd()
            if not os.getenv("DIR_ROOT_DIR")
            else os.getenv("DIR_ROOT_DIR"),
        )
        trainer.fit(model)
    except KeyboardInterrupt:
        torch.save(model.state_dict(), "INTERRUPTED.pth")
        logging.info("Saved interrupt")
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)


if __name__ == "__main__":
    main()


utils/dataset.py
import logging
import os
from glob import glob
from os import listdir
from os.path import splitext

from PIL import Image
from torch.utils.data import Dataset
from torchvision.transforms import transforms

from utils.pre_processing import (
    ToTensor,
    Rescale,
    RandomFlip,
    RandomColorJitter,
    RandomNoise,
    MaskToClasses,
)


class OrthogonalPhotoDataset(Dataset):
    def __init__(
        self,
        data_path: str = "data/",
        image_suffix: str = "_x",
        mask_suffix: str = "_y",
        augmentation: bool = True,
        image_size: int = 256,
        **kwargs,
    ):
        """
        Parameters
        ----------
        imgs_dir : str
            Directory for images.
        masks_dir : str
            Directory for masks.
        image_suffix :str
            Suffix for images.
        mask_suffix :str
            Suffix for masks.
        augmentation :bool
            Augmentation bool for tran/test sets.
        image_size :int
            Size to resize images to.
        """
        self.image_size = image_size
        self.augmentation = augmentation
        self.imgs_dir = os.path.join(data_path, "images/")
        self.masks_dir = os.path.join(data_path, "masks/")
        self.mask_suffix = mask_suffix
        self.image_suffix = image_suffix
        self.ids = [
            splitext(file)[0]
            for file in listdir(self.imgs_dir)
            if not file.startswith(".")
        ]
        logging.info(f"Creating dataset with {len(self.ids)} examples")
        self.mapping = {0: 0, 255: 1}

    def set_augmentation(self, aug: bool = False):
        """
        Sets augmentation to true or false.

        Parameters
        ----------
        aug : bool
            Data Augmentation on or off.
        """
        self.augmentation = aug

    def __len__(self):
        return len(self.ids)

    def __getitem__(self, i):
        """
        # Get image and mask
        """
        idx = self.ids[i]
        img_file = glob(self.imgs_dir + idx + ".*")
        msk_file = glob(
            str(self.masks_dir + idx + ".*").replace(
                self.image_suffix, self.mask_suffix
            )
        )

        if len(msk_file) != 1:
            raise AssertionError(
                f"Either no mask or multiple masks found for the ID {idx}: {msk_file}"
            )
        if len(img_file) != 1:
            raise AssertionError(
                f"Either no image or multiple images found for the ID {idx}: {img_file}"
            )
        img_as_img = Image.open(img_file[0]).convert("RGB")
        msk_as_img = Image.open(msk_file[0]).convert("L")

        if img_as_img.size != msk_as_img.size:
            raise AssertionError(
                f"Image and mask {idx} should be the same size, but are {img_as_img.size} and {msk_as_img.size}"
            )

        # Data augmentation
        if self.augmentation:
            transform = transforms.Compose(
                [
                    Rescale(self.image_size),
                    RandomFlip(),
                    RandomColorJitter(brightness=0.5, contrast=0.5, saturation=1.0),
                    RandomNoise(),
                    MaskToClasses(self.mapping),
                    ToTensor(),
                ]
            )
        else:
            transform = transforms.Compose(
                [Rescale(self.image_size), MaskToClasses(self.mapping), ToTensor()]
            )

        sample = {"image": img_as_img, "mask": msk_as_img}

        tranformed_sample = transform(sample)

        return tranformed_sample


unet/unet_modules.py
import torch
import torch.nn as nn
import torch.nn.functional as f


class DoubleConvolution(nn.Module):
    """
    Class used to initialize the conv 3x3, ReLu step.
    """

    def __init__(self, in_channels: int, out_channels: int, mid_channels: int = None):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        mid_channels : int
            Number if mid-layer channels
        """
        super().__init__()
        if not mid_channels:
            mid_channels = out_channels
        self.double_conv = nn.Sequential(
            nn.Conv2d(in_channels, mid_channels, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(mid_channels),
            nn.ReLU(inplace=True),
            nn.Conv2d(mid_channels, out_channels, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(out_channels),
            nn.ReLU(inplace=True),
        )

    def forward(self, x):
        """
        Preforms and returns results from the conv 3x3, ReLu step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.double_conv(x)


class Down(nn.Module):
    """
    Class used to initialize the max pool 2x2 step.
    """

    def __init__(self, in_channels: int, out_channels: int):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        """
        super().__init__()
        self.maxpool_conv = nn.Sequential(
            nn.MaxPool2d(2), DoubleConvolution(in_channels, out_channels)
        )

    def forward(self, x):
        """
        Preforms and returns results from the max pool 2x2 step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.maxpool_conv(x)


class Up(nn.Module):
    """
    Class used to initialize the up-conv 2x2 step.
    """

    def __init__(self, in_channels: int, out_channels: int, bilinear: bool = True):
        super().__init__()
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        bilinear : bool
            Bilinear interpolation in upsampling(default)
        """

        if bilinear:
            self.up = nn.Upsample(scale_factor=2, mode="bilinear", align_corners=True)
            self.conv = DoubleConvolution(in_channels, out_channels, in_channels // 2)
        else:
            self.up = nn.ConvTranspose2d(
                in_channels, in_channels // 2, kernel_size=2, stride=2
            )
            self.conv = DoubleConvolution(in_channels, out_channels)

    def forward(self, x1, x2):
        """
        Preforms and returns results from the up-conv 2x2 step.

        Parameters
        ----------
        x1 : torch.tensor
            From
        x2 : torch.tensor
            To
        """
        x1 = self.up(x1)
        diff_y = x2.size()[2] - x1.size()[2]
        diff_x = x2.size()[3] - x1.size()[3]

        x1 = f.pad(
            x1, [diff_x // 2, diff_x - diff_x // 2, diff_y // 2, diff_y - diff_y // 2]
        )
        x = torch.cat([x2, x1], dim=1)
        return self.conv(x)


class OutConvolution(nn.Module):
    """
    Class used to initialize the conv 1x1 step.
    """

    def __init__(self, in_channels: int, out_channels: int):
        """

        Parameters
        ----------
        in_channels : int
            Number of input channels
        out_channels : int
            Number of output channels
        """
        super(OutConvolution, self).__init__()
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size=1)

    def forward(self, x):
        """
        Preforms and returns results from the conv 1x1 step.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        return self.conv(x)


unet/unet_model.py
import copy
import os
from argparse import ArgumentParser
from random import randint

import kornia
import pytorch_lightning as pl
import torch
import torch.nn as nn
import torchvision
from pytorch_lightning.metrics.functional import f1_score, recall, precision
from torch.optim.lr_scheduler import ReduceLROnPlateau
from torch.utils.data import random_split, DataLoader

from unet.unet_modules import DoubleConvolution, Down, Up, OutConvolution
from utils.dataset import OrthogonalPhotoDataset


class UNet(pl.LightningModule):
    """
    Basic U-net structure as described in : O. Ronneberger, P. Fischer, and T. Brox, “U-net: Convolutional networks
    for biomedical image segmentation.” 2015.
    """

    def __init__(
        self,
        n_channels: int = 3,
        n_classes: int = 1,
        bilinear=True,
        lr: float = 0.0001,
        **kwargs,
    ):
        """

        Parameters
        ----------
        n_channels : int
            Number of channels in input-data.
        n_classes : int
            Number of classes to segment.
        bilinear : bool
            Bilinear interpolation in upsampling(default)
        """
        super().__init__()
        self.lr = lr
        self.save_hyperparameters()

        self.example_input_array = torch.randn(
            self.hparams.batch_size,
            n_channels,
            self.hparams.image_size,
            self.hparams.image_size,
        )

        if self.hparams.n_classes > 1:
            self.criterion = nn.CrossEntropyLoss()
        else:
            self.criterion = nn.BCEWithLogitsLoss()

        factor = 2 if bilinear else 1

        self.in_conv = DoubleConvolution(n_channels, 64)
        self.down_conv_1 = Down(64, 128)
        self.down_conv_2 = Down(128, 256)
        self.down_conv_3 = Down(256, 512)
        self.down_conv_4 = Down(512, 1024 // factor)

        self.up_conv_1 = Up(1024, 512 // factor, bilinear)
        self.up_conv_2 = Up(512, 256 // factor, bilinear)
        self.up_conv_3 = Up(256, 128 // factor, bilinear)
        self.up_conv_4 = Up(128, 64, bilinear)
        self.out_conv = OutConvolution(64, n_classes)

    def forward(self, x):
        """
        Feed-forward function.

        Parameters
        ----------
        x : torch.tensor
            Input data
        """
        x1 = self.in_conv(x)
        x2 = self.down_conv_1(x1)
        x3 = self.down_conv_2(x2)
        x4 = self.down_conv_3(x3)
        x5 = self.down_conv_4(x4)

        x = self.up_conv_1(x5, x4)
        x = self.up_conv_2(x, x3)
        x = self.up_conv_3(x, x2)
        x = self.up_conv_4(x, x1)
        out = self.out_conv(x)

        return out

    def setup(self, stage):
        dataset = OrthogonalPhotoDataset(**self.hparams)
        test_dataset = copy.deepcopy(OrthogonalPhotoDataset(**self.hparams))
        test_amount = int(len(dataset) * self.hparams.test_percent / 100)
        val_amount = int(len(dataset) * self.hparams.val_percent / 100)
        train_amount = len(dataset) - val_amount - test_amount

        # Development-mode
        if not os.getenv("PROD"):
            # Fix the generator for reproducible results
            self.train_set, self.val_set, self.test_set = random_split(
                dataset,
                [train_amount, val_amount, test_amount],
                generator=torch.Generator().manual_seed(42),
            )
        # Production-mode
        else:
            # Random generator
            self.train_set, self.val_set, self.test_set = random_split(
                dataset, [train_amount, val_amount, test_amount]
            )

        # Disable augmentation for validation set.
        self.val_set.dataset = test_dataset
        self.test_set.dataset = test_dataset
        self.val_set.dataset.set_augmentation(False)
        self.test_set.dataset.set_augmentation(False)

    def train_dataloader(self):
        return DataLoader(
            self.train_set,
            batch_size=self.hparams.batch_size,
            shuffle=True,
            num_workers=int(os.getenv("WORKERS", 4)),
            pin_memory=True,
        )

    def val_dataloader(self):
        return DataLoader(
            self.val_set,
            batch_size=self.hparams.batch_size,
            shuffle=False,
            num_workers=int(os.getenv("WORKERS", 4)),
            pin_memory=True,
            drop_last=True,
        )

    def test_dataloader(self):
        return DataLoader(
            self.test_set,
            batch_size=self.hparams.batch_size,
            shuffle=False,
            num_workers=int(os.getenv("WORKERS", 4)),
            pin_memory=True,
            drop_last=True,
        )

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(
            self.parameters(), lr=(self.lr or self.learning_rate)
        )
        lr_scheduler = ReduceLROnPlateau(optimizer, "min")
        scheduler = {
            "scheduler": lr_scheduler,
            "reduce_on_plateau": True,
            # val_checkpoint_on is val_loss passed in as checkpoint_on
            "monitor": "val_checkpoint_on",
            "patience": 5,
            "mode": "min",
            "factor": 0.1,
            "verbose": True,
            "min_lr": 1e-8,
        }
        return [optimizer], [scheduler]

    def loss_funciton(self, input, target):
        return self.criterion(input, target)

    def logg_images(self, images, masks, masks_pred, step: str = ""):
        rand_idx = randint(0, self.hparams.batch_size - 1)
        onehot = torch.sigmoid(masks_pred[rand_idx]) > 0.5
        for tag, value in self.named_parameters():
            tag = tag.replace(".", "/")
            self.logger.experiment.add_histogram(tag, value, self.current_epoch)
        mask_grid = torchvision.utils.make_grid([masks[rand_idx], onehot], nrow=2)
        self.logger.experiment.add_image(
            "{} - Target vs Predicted".format(step), mask_grid, self.current_epoch
        )
        alpha = 0.5
        image_grid = torchvision.utils.make_grid(
            [
                images[rand_idx],
                torch.clamp(
                    kornia.enhance.add_weighted(
                        src1=images[rand_idx],
                        alpha=1.0,
                        src2=onehot,
                        beta=alpha,
                        gamma=0.0,
                    ),
                    max=1.0,
                ),
            ]
        )
        self.logger.experiment.add_image(
            "{} - Image vs Predicted".format(step), image_grid, self.current_epoch
        )

    def training_step(self, batch, batch_idx):
        images, masks = batch["image"], batch["mask"]
        if images.shape[1] != self.hparams.n_channels:
            raise AssertionError(
                f"Network has been defined with {self.hparams.n_channels} input channels, "
                f"but loaded images have {images.shape[1]} channels. Please check that "
                "the images are loaded correctly."
            )

        masks = (
            masks.type(torch.float32)
            if self.hparams.n_classes == 1
            else masks.type(torch.long)
        )

        masks_pred = self(images)  # Forward pass
        loss = self.loss_funciton(masks_pred, masks)
        result = pl.TrainResult(minimize=loss)
        result.log("train_loss", loss, sync_dist=True)
        if batch_idx == 0:
            self.logg_images(images, masks, masks_pred, "TRAIN")
        pred = (torch.sigmoid(masks_pred) > 0.5).float()
        f1 = f1_score(pred, masks, self.hparams.n_classes + 1)
        rec = recall(pred, masks, self.hparams.n_classes + 1)
        pres = precision(pred, masks, self.hparams.n_classes + 1)
        result.log("train_f1", f1, on_epoch=True)
        result.log("train_recall", rec, on_epoch=True)
        result.log("train_precision", pres, on_epoch=True)

        return result

    def validation_step(self, batch, batch_idx):
        images, masks = batch["image"], batch["mask"]
        if images.shape[1] != self.hparams.n_channels:
            raise AssertionError(
                f"Network has been defined with {self.n_channels} input channels, "
                f"but loaded images have {images.shape[1]} channels. Please check that "
                "the images are loaded correctly."
            )

        masks = (
            masks.type(torch.float32)
            if self.hparams.n_classes == 1
            else masks.type(torch.long)
        )

        masks_pred = self(images)  # Forward pass
        loss = self.loss_funciton(masks_pred, masks)
        result = pl.EvalResult(loss, checkpoint_on=loss)
        result.log("val_loss", loss, sync_dist=True)
        if batch_idx == 0:
            self.logg_images(images, masks, masks_pred, "VAL")
        pred = (torch.sigmoid(masks_pred) > 0.5).float()
        f1 = f1_score(pred, masks, self.hparams.n_classes + 1)
        rec = recall(pred, masks, self.hparams.n_classes + 1)
        pres = precision(pred, masks, self.hparams.n_classes + 1)
        result.log("val_f1", f1, on_epoch=True)
        result.log("val_recall", rec, on_epoch=True)
        result.log("val_precision", pres, on_epoch=True)

        return result

    def test_step(self, batch, batch_idx):
        images, masks = batch["image"], batch["mask"]
        if images.shape[1] != self.hparams.n_channels:
            raise AssertionError(
                f"Network has been defined with {self.n_channels} input channels, "
                f"but loaded images have {images.shape[1]} channels. Please check that "
                "the images are loaded correctly."
            )

        masks = (
            masks.type(torch.float32)
            if self.hparams.n_classes == 1
            else masks.type(torch.long)
        )

        masks_pred = self(images)  # Forward pass
        loss = self.loss_funciton(masks_pred, masks)
        result = pl.EvalResult(loss, checkpoint_on=loss)
        result.log("test_loss", loss, on_step=True, on_epoch=True, sync_dist=True)
        self.logg_images(images, masks, masks_pred, "TEST")
        pred = (torch.sigmoid(masks_pred) > 0.5).float()
        f1 = f1_score(pred, masks, self.hparams.n_classes + 1)
        rec = recall(pred, masks, self.hparams.n_classes + 1)
        pres = precision(pred, masks, self.hparams.n_classes + 1)
        result.log("test_f1", f1, on_epoch=True)
        result.log("test_recall", rec, on_epoch=True)
        result.log("test_precision", pres, on_epoch=True)

        return result

    @staticmethod
    def add_model_specific_args(parent_parser):
        parser = ArgumentParser(parents=[parent_parser], add_help=False)
        parser.add_argument(
            "-ch",
            "--n-channels",
            dest="n_channels",
            type=int,
            metavar="NCHS",
            default=3 if not os.getenv("N_CHANNELS") else os.getenv("N_CHANNELS"),
            help="Number of channels in input.",
        )
        parser.add_argument(
            "-cl",
            "--n-classes",
            dest="n_classes",
            type=int,
            metavar="NCLS",
            default=1 if not os.getenv("N_CLASSES") else os.getenv("N_CLASSES"),
            help="Number of classes to classify.",
        )
        parser.add_argument(
            "-lr",
            "--learning-rate",
            dest="lr",
            type=float,
            metavar="LR",
            default=0.0001 if not os.getenv("LRN_RATE") else os.getenv("LRN_RATE"),
            help="Number of classes to classify.",
        )
        parser.add_argument(
            "-b",
            "--bilinear",
            dest="bilinear",
            type=bool,
            default=True if not os.getenv("BILINEAR") else os.getenv("BILINEAR"),
            help="To use bilinear up-scaling i in the model.",
        )
        parser.add_argument(
            "-is",
            "--image-size",
            dest="image_size",
            type=int,
            metavar="IS",
            default=512 if not os.getenv("IMG_SIZE") else os.getenv("IMG_SIZE"),
            help="Image height value for resizing.",
        )
        parser.add_argument(
            "-valp",
            "--validation-percent",
            metavar="VALP",
            dest="val_percent",
            type=float,
            default=10.0 if not os.getenv("VAL_PERC") else os.getenv("VAL_PERC"),
            help="How much of dataset to be used as validation set.",
        )
        parser.add_argument(
            "-tesp",
            "--testing-percent",
            metavar="TESP",
            dest="test_percent",
            type=float,
            default=10.0 if not os.getenv("TEST_PERC") else os.getenv("TEST_PERC"),
            help="How much of dataset to be used as testing set.",
        )
        parser.add_argument(
            "-bs",
            "--batch-size",
            dest="batch_size",
            metavar="BS",
            type=int,
            default=1 if not os.getenv("BATCH_SIZE") else os.getenv("BATCH_SIZE"),
            help="Batch size",
        )
        parser.add_argument(
            "-dp",
            "--data-path",
            metavar="DP",
            dest="data_path",
            type=str,
            default="data" if not os.getenv("DIR_DATA") else os.getenv("DIR_DATA"),
            help="Load model from a .pth file",
        )

        return parser
